src/node_api.cc

```diff
@@ -3287,3 +3349,55 @@ napi_status napi_cancel_async_work(napi_env env, napi_async_work work) {
 
   return napi_clear_last_error(env);
 }
+
+NAPI_EXTERN napi_status napi_create_promise(napi_env env,
+                                            napi_value* deferred,
+                                            napi_value* promise) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(env, deferred);
+  CHECK_ARG(env, promise);
+
+  auto maybe = v8::Promise::Resolver::New(env->isolate->GetCurrentContext());
+  CHECK_MAYBE_EMPTY(env, maybe, napi_generic_failure);
+
+  auto v8_resolver = maybe.ToLocalChecked();
+
+  // Stick the V8 resolver into an empty shell object.
+  v8::Local<v8::Object> wrapper;
+  napi_status status = v8impl::EnvWrapperObject(env,
+                                                &env->deferred_template,
+                                                v8_resolver,
+                                                v8impl::napi_deferred_name,
+                                                &wrapper);
+  if (status != napi_ok) {
+    return napi_set_last_error(env, status);
+  }
+
+  *deferred = v8impl::JsValueFromV8LocalValue(wrapper);
+  *promise = v8impl::JsValueFromV8LocalValue(v8_resolver->GetPromise());
+  return GET_RETURN_STATUS(env);
+}
+
+NAPI_EXTERN napi_status napi_fulfill_deferred(napi_env env,
```

Comment:
By the way, I find `fulfill` odd here, because it *is* resolving and not necessarily fulfilling the promise? Why not use the more accurate name, especially if the parameter names, the internal methods for it all use “resolve” as well?